mlir/include/mlir/Dialect/Tensor/IR/Tensor.h

```diff
@@ -176,6 +176,9 @@ void populateFoldConstantExtractSlicePatterns(
           return false;
         });
 
+/// Patterns to fold inserts into a constant into a new constant.
+void populateFoldInsertAfterConstant(RewritePatternSet &patterns);
```

Comment:
I'm not comfortable with dedicated API exposing single patterns, that does not seem super principled to me in general unfortunately. We should group things in optimization passes that are coherent and consistent.

Otherwise this leads to a zoo of APIs that are just completely ad-hoc and not really usable IMO.

Reply:
Would it be OK to expose the single patterns in the header and have downstream users use the pattern? Grouping them in a pass means choosing some heuristic generally for users for when to apply this pattern, or I could make a tensor size option for it.

Reply:
> Would it be OK to expose the single patterns in the header and have downstream users use the pattern?

Isn't this was you're doing here? This seems like what I'm objecting to actually: ad-hoc APIs exposing a zoo of patterns that downstream have to somehow managed to package together consistently?

Reply:
OK - fair enough, exposing a populate*Pattern method and the pattern itself.

There's a similar pass in memref ops like `-fold-memref-alias-ops` (the constant -> insert one is different, but the extract from tensor.collapse_shape falls in that bucket). All the patterns I want are some kind of tensor folding operations.

I'm also lowering StableHLO -> linalg and want to add a few canonicalizations for
* collapsing a constant tensor folds to a collapsed constant
* broadcasting a constant tensor folds to a broadcasted constant
* a linalg.fill of a scalar constant folds to a tensor constant

Maybe a dedicated tensor folder pass would be good to start? `-fold-tensor-ops`?

Reply:
> ad-hoc APIs exposing a zoo of patterns that downstream have to somehow managed to package together consistently

It sounds like packaging them in a pass that can later be enhanced with a cost model is a good approach.

Reply:
I had an internship proposal to work on a module/function level constant folding pass with a pluggable cost-model, but the internship got canceled unfortunately :(
(It's non-trivial to design, which is the kind of thing I like having an intern explore for some time)

Reply:
Maybe `MergeConsecutiveInsertExtractSlicePatterns.cpp` could be generalized to include this pattern.

Reply:
It's probably safer to remove the pattern for now, and maybe think about introducing a "tensor-constant-folding" pass separately. I suspect such pass would want options about a maximum size for folding a constant or other option for the strategy to adopt.